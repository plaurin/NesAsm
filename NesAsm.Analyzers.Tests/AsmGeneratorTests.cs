using ApprovalTests.Reporters;

namespace NesAsm.Analyzers.Tests;

[UseReporter(typeof(VisualStudioReporter))]
public class AsmGeneratorTests
{
    [Fact]
    public void TestExampleScript()
    {
        var generator = new AsmGenerator();
        var content = File.ReadAllText($@"..\..\..\..\NesAsm.Example\HiloWorld.cs");

        GeneratorTestUtilities.TestGenerator(generator, content);
    }

    [Fact]
    public void TestSimpleScript()
    {
        TestScript("SimpleScript.cs");
    }

    [Fact]
    public void TestMultiProcScript()
    {
        TestScript("MultiProcScript.cs");
    }

    [Fact]
    public void TestJumpSubroutineScript()
    {
        TestScript("JumpSubroutineScript.cs");
    }

    [Fact]
    public void TestCharsScript()
    {
        TestScript("CharsScript.cs");
    }

    [Fact]
    public void TestBranchScript()
    {
        TestScript("BranchScript.cs");
    }

    [Fact]
    public void TestParametersScript()
    {
        TestScript("ParametersScript.cs");
    }

    [Fact]
    public void TestSubroutineResultScript()
    {
        TestScript("SubroutineResultScript.cs");
    }

    [Fact]
    public void TestSubroutineResultsScript()
    {
        TestScript("SubroutineResultsScript.cs");
    }

    [Fact]
    public void TestCallingOtherScript()
    {
        TestScript("CallingOtherScript.cs");
    }

    [Fact]
    public void TestDataScript()
    {
        TestScript("DataScript.cs");
    }

    [Fact]
    public void TestAllInstructionsScript()
    {
        TestScript("AllInstructionsScript.cs");
    }

    private static void TestScript(string filename)
    {
        var generator = new AsmGenerator();
        var content = File.ReadAllText($@"..\..\..\TestFiles\{filename}");

        GeneratorTestUtilities.TestGenerator(generator, content);
    }
}
