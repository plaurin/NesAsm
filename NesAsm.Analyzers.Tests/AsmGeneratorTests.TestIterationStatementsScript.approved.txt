- Main diagnostics:
(55,9): error NA0013: Only X or Y register supported for loop variable but found Z
(60,9): error NA0015: Loop cannot use local variable, remove type declaration int
(65,9): error NA0014: Loop should only use 1 register
(76,9): warning NA0008: Instruction is not currently supported: do
(76,9): error NA0005: Invalid format true for type byte or ushort
(52,5): error NA0016: Loop not closed, ensure that all for loop block use bracket \{ \}
- SyntaxTrees Count: 2
- Output diagnostics:
- RunResult.GeneratedTrees Count: 1
- RunResults diagnostics:
(55,9): error NA0013: Only X or Y register supported for loop variable but found Z
(60,9): error NA0015: Loop cannot use local variable, remove type declaration int
(65,9): error NA0014: Loop should only use 1 register
(76,9): warning NA0008: Instruction is not currently supported: do
(76,9): error NA0005: Invalid format true for type byte or ushort
(52,5): error NA0016: Loop not closed, ensure that all for loop block use bracket \{ \}
- RunResult[0].Generator: Microsoft.CodeAnalysis.IncrementalGeneratorWrapper
- RunResult[0] diagnostics:
(55,9): error NA0013: Only X or Y register supported for loop variable but found Z
(60,9): error NA0015: Loop cannot use local variable, remove type declaration int
(65,9): error NA0014: Loop should only use 1 register
(76,9): warning NA0008: Instruction is not currently supported: do
(76,9): error NA0005: Invalid format true for type byte or ushort
(52,5): error NA0016: Loop not closed, ensure that all for loop block use bracket \{ \}
- RunResult[0].GeneratedSources Count: 1
- RunResult[0].Exception: 
- RunResult[0].GeneratedSources[0].HintName: Asm.IterationStatementsScript.cs
- RunResult[0].GeneratedSources[0].SourceText:
/* Generator Asm code in file Asm.IterationStatementsScript.s
; Auto generated code using the NesAsm project
.scope IterationStatementsScript

  .proc main
    ; for - using X or Y register

    ldx #0
  @loop1_on_X:
    lda data, x

    sta $2007

    inx
    cpx #10
    bne @loop1_on_X


    ldy #0
  @loop2_on_Y:
    lda data, y

    iny
    cpy #10
    bne @loop2_on_Y

    ; for - nested loop

    ldx #0
  @loop3_on_X:

    ldy #0
  @loop4_on_Y:
    lda data, x
    lda data, y

    iny
    cpy #4
    bne @loop4_on_Y

    inx
    cpx #4
    bne @loop3_on_X

    ; TODO while condition

    ; while true - infinite loop
  @endless_loop:
    lda #42
    jmp @endless_loop

    rts
  .endproc

.segment "CODE"

data:
  .byte 1
  .byte 2
  .byte 3
  .byte 4
  .byte 5
  .byte 6
  .byte 7
  .byte 8

  .byte 9
  .byte 10

  .proc invalidFor
    ; NA0013 - Should be X or Y only

    ; NA0015 - Should not declare new local variable X

    ; NA0014 - Should not mix X and Y

    ; NA0016 - Should not use the same register in nester loop

    ldx #0
  @loop1_on_X:

    ldx #0
  @loop2_on_X:

    inx
    cpx #10
    bne @loop2_on_X

    ; NA0008 - do is not supported

    rts
  .endproc

.endscope

.include "wrapper.s"

*/
